test/unit/function_resolver_test.rb

```diff
@@ -239,291 +242,378 @@ class FunctionResolverTest < Test::Unit::TestCase
+    f = FunctionResolver.new('hours', [{ type: :integer, value: -24 }])
+    f.expects(:current_time).returns(test_time)
     f.validate
     assert !f.errors?, "Errors #{f.errors.inspect}"
     value = f.call
-    
+
     assert_equal :datetime, value[:type]
     d = DateTime.parse(value[:value])
-    assert_equal dt.year, d.year
-    assert_equal dt.month - 1, d.month
+    assert_equal test_time.year, d.year
+    assert_equal test_time.month - 1, d.month
     assert_equal 31, d.mday
-    assert_equal dt.hour, d.hour
-    assert_equal dt.min, d.min
+    assert_equal test_time.hour, d.hour
+    assert_equal test_time.min, d.min
 
     # Drop back one full year's worth of hours.
-    f = FunctionResolver.new('hours', [{:type=>:integer, :value => -8760}])
+    f = FunctionResolver.new('hours', [{ type: :integer, value: -8760 }])
+    f.expects(:current_time).returns(test_time)
     f.validate
     assert !f.errors?, "Errors #{f.errors.inspect}"
     value = f.call
-    
+
     assert_equal :datetime, value[:type]
     d = DateTime.parse(value[:value])
-    assert_equal dt.year - 1, d.year
-    assert_equal dt.month, d.month
-    assert_equal dt.mday, d.mday
-    assert_equal dt.hour, d.hour
-    assert_equal dt.min, d.min
+    assert_equal test_time.year - 1, d.year
+    assert_equal test_time.month, d.month
+    assert_equal test_time.mday, d.mday
+    assert_equal test_time.hour, d.hour
+    assert_equal test_time.min, d.min
   end
 
-  test "days()" do
-    d = Date.new(2012,10,20)
-    Date.expects(:today).returns(d)
-    dt = DateTime.new(d.year, d.month,d.day, 0,0,0, DateTime.now.offset)
-    start = Time.parse(dt.to_s)
-    f = FunctionResolver.new('days', [{:type=>:integer, :value =>7}])
+  test 'days()' do
+    test_date = Date.new(2012, 10, 20) # Sat, 20 Oct 2012 00:00:00 GMT
+    f = FunctionResolver.new('days', [{ type: :integer, value: 7 }])
+    f.expects(:current_date).returns(test_date)
     f.validate
     assert !f.errors?, "Errors #{f.errors.inspect}"
     value = f.call
     assert_equal :date, value[:type]
-    test_time = Time.parse(value[:value])
-    assert (615000 > test_time - start && 600000 < test_time - start), "Time range off by more than five seconds #{test_time - start} '#{test_time} - #{start}'"
+    assert_equal '2012-10-27', value[:value]
+  end
+
+  test 'weekdays()' do
+    friday = Date.new(2012, 10, 19)
+    saturday = Date.new(2012, 10, 20)
+    sunday = Date.new(2012, 10, 21)
+    monday = Date.new(2012, 10, 22)
+    {
+      friday => [
+        [-5, '2012-10-12'],
+        [-4, '2012-10-15'],
+        [-1, '2012-10-18'],
+        [0, '2012-10-19'],
+        [1, '2012-10-22'],
+        [2, '2012-10-23'],
+        [5, '2012-10-26'],
+        [6, '2012-10-29'],
+        [7, '2012-10-30'],
+        [31, '2012-12-03']
+      ],
+      saturday => [
+        [-6, '2012-10-12'],
+        [-5, '2012-10-15'],
+        [-1, '2012-10-19'],
+        [0, '2012-10-22'],
```

Comment:
Oh right, I did make a gut decision here on what to do with zero on a weekend. My take is to move it to the Monday. I've compared this with some online calculators, and they just didn't support zero, which I would be OK with as well.